presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java

```diff
@@ -553,7 +558,17 @@ public PlanOptimizers(
                         estimatedExchangesCostCalculator,
                         ImmutableSet.of(new SimplifyCountOverConstant(metadata))),
                 new LimitPushDown(), // Run LimitPushDown before WindowFilterPushDown
-                new WindowFilterPushDown(metadata), // This must run after PredicatePushDown and LimitPushDown so that it squashes any successive filter nodes and limits
+                new IterativeOptimizer(
+                        ruleStats,
+                        statsCalculator,
+                        estimatedExchangesCostCalculator,
+                        ImmutableList.of(new WindowFilterPushDown(metadata)),
+                        ImmutableSet.of(
+                                new PushdownLimitThroughRowNumber(),
+                                new PushdownLimitThroughWindow(metadata),
+                                new PushdownFilterThroughRowNumber(metadata),
+                                new PushdownFilterThroughWindow(metadata),
+                                new ReplaceWindowWithRowNumber(metadata))),
```

Comment:
Let's add a session property to control this. You can use the newly introduced `Predicate<Session> useLegacyRules` argument to IterativeOptimizer to determine whether the legacy or new rules should be used based on the session property.